File: jiuwenswarm/server/live_voice/product_composition_contract.py
# Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.

"""Pure Gate-0 contract for Live Voice product composition.

This module is intentionally not a composition root.  It validates bounded route
facts and maps the older Web-shell diagnostic vocabulary into the product truth
vocabulary without importing, constructing, starting, or calling any Adapter.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from enum import StrEnum


PRODUCT_COMPOSITION_CONTRACT_VERSION = "live-voice.product-composition.gate0.v1"


class ProductCompositionContractViolation(ValueError):
    """Raised when a route fact would overstate composition truth."""


class ProductRouteTruth(StrEnum):
    FORMAL = "formal"
    FALLBACK = "fallback"
    DEMO_SUBSTITUTE = "demo_substitute"
    UNAVAILABLE = "unavailable"
    DISABLED = "disabled"


class ProductSegment(StrEnum):
    AUTHORITY = "authority"
    P1_SPEECH_MEDIA = "p1.speech_media"
    P2_AGENT_INTERACTION = "p2.agent_interaction"
    P3_QUERY = "p3.query"
    P3_CONTROL = "p3.control"
    P3_PROGRESS = "p3.progress"
    BROWSER_AUDIO = "browser.audio"
    OBSERVABILITY = "observability"


class ProductRouteReason(StrEnum):
    FORMAL_ROUTE_OBSERVED = "FORMAL_ROUTE_OBSERVED"
    EXPLICIT_FALLBACK_ACTIVE = "EXPLICIT_FALLBACK_ACTIVE"
    D047_DEMO_SUBSTITUTE_ACTIVE = "D047_DEMO_SUBSTITUTE_ACTIVE"
    FEATURE_DISABLED = "FEATURE_DISABLED"
    ADAPTER_NOT_REGISTERED = "ADAPTER_NOT_REGISTERED"
    REQUESTED_ROUTE_UNAVAILABLE = "REQUESTED_ROUTE_UNAVAILABLE"
    FORMAL_ACTIVATION_EVIDENCE_MISSING = "FORMAL_ACTIVATION_EVIDENCE_MISSING"
    TRUSTED_AUTHORITY_UNAVAILABLE = "TRUSTED_AUTHORITY_UNAVAILABLE"
    SPEECH_AUTHORIZATION_UNAVAILABLE = "SPEECH_AUTHORIZATION_UNAVAILABLE"
    MEDIA_AUTHORITY_UNAVAILABLE = "MEDIA_AUTHORITY_UNAVAILABLE"
    MEDIA_PROVIDER_UNAVAILABLE = "MEDIA_PROVIDER_UNAVAILABLE"
    MEDIA_TRANSPORT_UNAVAILABLE = "MEDIA_TRANSPORT_UNAVAILABLE"
    MEDIA_LOGGER_ZERO_PERSISTENCE_UNPROVEN = "MEDIA_LOGGER_ZERO_PERSISTENCE_UNPROVEN"
    P2_AUTHORITY_UNAVAILABLE = "P2_AUTHORITY_UNAVAILABLE"
    P2_RUNTIME_UNAVAILABLE = "P2_RUNTIME_UNAVAILABLE"
    P2_NOTIFICATION_BACKPRESSURE_UNRESOLVED = "P2_NOTIFICATION_BACKPRESSURE_UNRESOLVED"
    P3_QUERY_AUTHORITY_UNAVAILABLE = "P3_QUERY_AUTHORITY_UNAVAILABLE"
    P3_CONFIRMATION_ISSUER_UNAVAILABLE = "P3_CONFIRMATION_ISSUER_UNAVAILABLE"
    TASK_PROGRESS_AUTHORITY_HANDOFF_UNAVAILABLE = (
        "TASK_PROGRESS_AUTHORITY_HANDOFF_UNAVAILABLE"
    )
    BROWSER_AUDIO_UNAVAILABLE = "BROWSER_AUDIO_UNAVAILABLE"
    OBSERVABILITY_CONSUMER_UNAVAILABLE = "OBSERVABILITY_CONSUMER_UNAVAILABLE"
    ACTIVATION_FAILED = "ACTIVATION_FAILED"


class ProductEvidenceId(StrEnum):
    GATE0_CONTRACT_ONLY = "GATE0_CONTRACT_ONLY"
    FEATURE_FLAG_OFF = "FEATURE_FLAG_OFF"
    TRUSTED_AUTHORITY_RESOLVED = "TRUSTED_AUTHORITY_RESOLVED"
    FORMAL_ACTIVATION_LEASE_OPEN = "FORMAL_ACTIVATION_LEASE_OPEN"
    RUNTIME_PATH_OBSERVED = "RUNTIME_PATH_OBSERVED"
    P2_NOTIFICATION_BACKPRESSURE_CLOSED = "P2_NOTIFICATION_BACKPRESSURE_CLOSED"
    MEDIA_LOGGER_ZERO_PERSISTENCE_VERIFIED = "MEDIA_LOGGER_ZERO_PERSISTENCE_VERIFIED"
    FALLBACK_ROUTE_SELECTED = "FALLBACK_ROUTE_SELECTED"
    D047_LEGACY_ROUTE_SELECTED = "D047_LEGACY_ROUTE_SELECTED"
    FORMAL_BATCH_SPEECH_FOUNDATION = "FORMAL_BATCH_SPEECH_FOUNDATION"
    D059_AGENT_CR_FOUNDATION = "D059_AGENT_CR_FOUNDATION"
    P3_QUERY_FOUNDATION = "P3_QUERY_FOUNDATION"
    TASK_EVENT_LIVE_ONLY = "TASK_EVENT_LIVE_ONLY"
    PROGRESS_ARBITER_CONTIGUOUS_SEQUENCE_REQUIRED = (
        "PROGRESS_ARBITER_CONTIGUOUS_SEQUENCE_REQUIRED"
    )
    BROWSER_AUDIO_FOUNDATION = "BROWSER_AUDIO_FOUNDATION"
    OBSERVABILITY_FOUNDATION = "OBSERVABILITY_FOUNDATION"
    P2_NOTIFICATION_QUEUE_BLOCKING_RISK = "P2_NOTIFICATION_QUEUE_BLOCKING_RISK"
    DEV_AUDIO_LOG_PERSISTENCE_RISK = "DEV_AUDIO_LOG_PERSISTENCE_RISK"
    PREINTEGRATION_STATIC_AUDIT_ONLY = "PREINTEGRATION_STATIC_AUDIT_ONLY"
    PACKAGE_CONTRACT_ONLY = "PACKAGE_CONTRACT_ONLY"
    NO_RUNTIME_EVIDENCE = "NO_RUNTIME_EVIDENCE"


_REASON_BY_TRUTH = {
    ProductRouteTruth.FORMAL: frozenset({ProductRouteReason.FORMAL_ROUTE_OBSERVED}),
    ProductRouteTruth.FALLBACK: frozenset(
        {ProductRouteReason.EXPLICIT_FALLBACK_ACTIVE}
    ),
    ProductRouteTruth.DEMO_SUBSTITUTE: frozenset(
        {ProductRouteReason.D047_DEMO_SUBSTITUTE_ACTIVE}
    ),
    ProductRouteTruth.DISABLED: frozenset({ProductRouteReason.FEATURE_DISABLED}),
    ProductRouteTruth.UNAVAILABLE: frozenset(
        reason
        for reason in ProductRouteReason
        if reason
        not in {
            ProductRouteReason.FORMAL_ROUTE_OBSERVED,
            ProductRouteReason.EXPLICIT_FALLBACK_ACTIVE,
            ProductRouteReason.D047_DEMO_SUBSTITUTE_ACTIVE,
            ProductRouteReason.FEATURE_DISABLED,
        }
    ),
}

_FORMAL_EVIDENCE = frozenset(
    {
        ProductEvidenceId.TRUSTED_AUTHORITY_RESOLVED,
        ProductEvidenceId.FORMAL_ACTIVATION_LEASE_OPEN,
        ProductEvidenceId.RUNTIME_PATH_OBSERVED,
    }
)

_FORMAL_STOP_CLOSURE_BY_SEGMENT = {
    ProductSegment.P1_SPEECH_MEDIA: (
        ProductEvidenceId.MEDIA_LOGGER_ZERO_PERSISTENCE_VERIFIED
    ),
    ProductSegment.P2_AGENT_INTERACTION: (
        ProductEvidenceId.P2_NOTIFICATION_BACKPRESSURE_CLOSED
    ),
}

_FORBIDDEN_EVIDENCE_BY_TRUTH = {
    ProductRouteTruth.FORMAL: frozenset(
        {
            ProductEvidenceId.GATE0_CONTRACT_ONLY,
            ProductEvidenceId.FEATURE_FLAG_OFF,
            ProductEvidenceId.FALLBACK_ROUTE_SELECTED,
            ProductEvidenceId.D047_LEGACY_ROUTE_SELECTED,
            ProductEvidenceId.P2_NOTIFICATION_QUEUE_BLOCKING_RISK,
            ProductEvidenceId.DEV_AUDIO_LOG_PERSISTENCE_RISK,
            ProductEvidenceId.NO_RUNTIME_EVIDENCE,
        }
    ),
    ProductRouteTruth.FALLBACK: frozenset(
        _FORMAL_EVIDENCE
        | {
            ProductEvidenceId.FEATURE_FLAG_OFF,
            ProductEvidenceId.D047_LEGACY_ROUTE_SELECTED,
        }
    ),
    ProductRouteTruth.DEMO_SUBSTITUTE: frozenset(
        _FORMAL_EVIDENCE
        | {
            ProductEvidenceId.FEATURE_FLAG_OFF,
            ProductEvidenceId.FALLBACK_ROUTE_SELECTED,
        }
    ),
    ProductRouteTruth.UNAVAILABLE: frozenset(
        {
            ProductEvidenceId.FEATURE_FLAG_OFF,
            ProductEvidenceId.FORMAL_ACTIVATION_LEASE_OPEN,
            ProductEvidenceId.RUNTIME_PATH_OBSERVED,
            ProductEvidenceId.FALLBACK_ROUTE_SELECTED,
            ProductEvidenceId.D047_LEGACY_ROUTE_SELECTED,
        }
    ),
    ProductRouteTruth.DISABLED: frozenset(
        evidence
        for evidence in ProductEvidenceId
        if evidence is not ProductEvidenceId.FEATURE_FLAG_OFF
    ),
}


@dataclass(frozen=True, slots=True)
class ProductRouteFact:
    segment: ProductSegment
    truth: ProductRouteTruth
    reason_id: ProductRouteReason
    evidence_ids: tuple[ProductEvidenceId, ...]
    formal_runtime_observed: bool = False

    def __post_init__(self) -> None:
        if not isinstance(self.segment, ProductSegment):
            raise ProductCompositionContractViolation(
                "segment must be a ProductSegment"
            )
        if not isinstance(self.truth, ProductRouteTruth):
            raise ProductCompositionContractViolation(
                "truth must be a ProductRouteTruth"
            )
        if not isinstance(self.reason_id, ProductRouteReason):
            raise ProductCompositionContractViolation(
                "reason_id must be a ProductRouteReason"
            )
        if self.reason_id not in _REASON_BY_TRUTH[self.truth]:
            raise ProductCompositionContractViolation(
                "reason_id is incompatible with route truth"
            )
        if type(self.evidence_ids) is not tuple or not self.evidence_ids:
            raise ProductCompositionContractViolation(
                "evidence_ids must be a non-empty tuple"
            )
        if any(
            not isinstance(evidence_id, ProductEvidenceId)
            for evidence_id in self.evidence_ids
        ):
            raise ProductCompositionContractViolation(
                "evidence_ids must contain ProductEvidenceId values"
            )
        if len(set(self.evidence_ids)) != len(self.evidence_ids):
            raise ProductCompositionContractViolation(
                "evidence_ids must not contain duplicates"
            )
        if type(self.formal_runtime_observed) is not bool:
            raise ProductCompositionContractViolation(
                "formal_runtime_observed must be a boolean"
            )

        evidence = frozenset(self.evidence_ids)
        if evidence & _FORBIDDEN_EVIDENCE_BY_TRUTH[self.truth]:
            raise ProductCompositionContractViolation(
                "route truth contains contradictory evidence"
            )
        if self.truth is ProductRouteTruth.FORMAL:
            if not self.formal_runtime_observed or not _FORMAL_EVIDENCE.issubset(
                evidence
            ):
                raise ProductCompositionContractViolation(
                    "formal truth requires trusted authority, an open activation "
                    "lease, and observed runtime-path evidence"
                )
            required_stop_closure = _FORMAL_STOP_CLOSURE_BY_SEGMENT.get(self.segment)
            if (
                required_stop_closure is not None
                and required_stop_closure not in evidence
            ):
                raise ProductCompositionContractViolation(
                    "formal segment is missing its affirmative activation-stop "
                    "closure evidence"
                )
        elif self.formal_runtime_observed:
            raise ProductCompositionContractViolation(
                "non-formal truth cannot claim formal runtime observation"
            )

        required_evidence = {
            ProductRouteTruth.FALLBACK: ProductEvidenceId.FALLBACK_ROUTE_SELECTED,
            ProductRouteTruth.DEMO_SUBSTITUTE: (
                ProductEvidenceId.D047_LEGACY_ROUTE_SELECTED
            ),
            ProductRouteTruth.DISABLED: ProductEvidenceId.FEATURE_FLAG_OFF,
        }.get(self.truth)
        if required_evidence is not None and required_evidence not in evidence:
            raise ProductCompositionContractViolation(
                "route truth is missing its required evidence identifier"
            )


@dataclass(frozen=True, slots=True)
class ProductCompositionManifest:
    contract_version: str
    enabled: bool
    routes: tuple[ProductRouteFact, ...]

    def __post_init__(self) -> None:
        if self.contract_version != PRODUCT_COMPOSITION_CONTRACT_VERSION:
            raise ProductCompositionContractViolation(
                "unsupported product composition contract version"
            )
        if type(self.enabled) is not bool:
            raise ProductCompositionContractViolation("enabled must be a boolean")
        if type(self.routes) is not tuple:
            raise ProductCompositionContractViolation("routes must be a tuple")
        if any(not isinstance(route, ProductRouteFact) for route in self.routes):
            raise ProductCompositionContractViolation(
                "routes must contain ProductRouteFact values"
            )
        if tuple(route.segment for route in self.routes) != tuple(ProductSegment):
            raise ProductCompositionContractViolation(
                "manifest routes must contain every segment in canonical order"
            )
        if not self.enabled and any(
            route.truth is not ProductRouteTruth.DISABLED for route in self.routes
        ):
            raise ProductCompositionContractViolation(
                "disabled manifest must report every segment disabled"
            )
        authority = self.routes[0]
        if (
            any(route.truth is ProductRouteTruth.FORMAL for route in self.routes[1:])
            and authority.truth is not ProductRouteTruth.FORMAL
        ):
            raise ProductCompositionContractViolation(
                "formal product segments require a formal authority segment"
            )


def _disabled_fact(segment: ProductSegment) -> ProductRouteFact:
    return ProductRouteFact(
        segment=segment,
        truth=ProductRouteTruth.DISABLED,
        reason_id=ProductRouteReason.FEATURE_DISABLED,
        evidence_ids=(ProductEvidenceId.FEATURE_FLAG_OFF,),
    )


def create_product_composition_manifest(
    *, enabled: bool, route_facts: Iterable[ProductRouteFact] | object = ()
) -> ProductCompositionManifest:
    """Build a pure manifest; feature-off returns before inspecting route facts."""

    if type(enabled) is not bool:
        raise ProductCompositionContractViolation("enabled must be a boolean")
    if not enabled:
        return ProductCompositionManifest(
            contract_version=PRODUCT_COMPOSITION_CONTRACT_VERSION,
            enabled=False,
            routes=tuple(_disabled_fact(segment) for segment in ProductSegment),
        )

    if not isinstance(route_facts, Iterable):
        raise ProductCompositionContractViolation("route_facts must be iterable")
    facts = tuple(route_facts)
    if any(not isinstance(fact, ProductRouteFact) for fact in facts):
        raise ProductCompositionContractViolation(
            "route_facts must contain ProductRouteFact values"
        )
    by_segment: dict[ProductSegment, ProductRouteFact] = {}
    for fact in facts:
        if fact.segment in by_segment:
            raise ProductCompositionContractViolation(
                "route_facts must not contain duplicate segments"
            )
        by_segment[fact.segment] = fact

    routes = tuple(
        by_segment.get(
            segment,
            ProductRouteFact(
                segment=segment,
                truth=ProductRouteTruth.UNAVAILABLE,
                reason_id=ProductRouteReason.ADAPTER_NOT_REGISTERED,
                evidence_ids=(ProductEvidenceId.GATE0_CONTRACT_ONLY,),
            ),
        )
        for segment in ProductSegment
    )
    return ProductCompositionManifest(
        contract_version=PRODUCT_COMPOSITION_CONTRACT_VERSION,
        enabled=True,
        routes=routes,
    )


def route_fact_from_integrated_shell(
    *,
    segment: ProductSegment,
    feature_enabled: bool,
    legacy_route_class: object,
    formal_runtime_observed: bool = False,
    formal_evidence_ids: tuple[ProductEvidenceId, ...] = (),
) -> ProductRouteFact:
    """Map existing shell diagnostics without upgrading manifests to formal truth."""

    if type(feature_enabled) is not bool:
        raise ProductCompositionContractViolation("feature_enabled must be a boolean")
    if not feature_enabled:
        return _disabled_fact(segment)
    if not isinstance(segment, ProductSegment):
        raise ProductCompositionContractViolation("segment must be a ProductSegment")
    if type(formal_runtime_observed) is not bool:
        raise ProductCompositionContractViolation(
            "formal_runtime_observed must be a boolean"
        )
    if type(legacy_route_class) is not str:
        raise ProductCompositionContractViolation("legacy_route_class must be a string")

    if legacy_route_class == "fallback":
        return ProductRouteFact(
            segment=segment,
            truth=ProductRouteTruth.FALLBACK,
            reason_id=ProductRouteReason.EXPLICIT_FALLBACK_ACTIVE,
            evidence_ids=(ProductEvidenceId.FALLBACK_ROUTE_SELECTED,),
        )
    if legacy_route_class == "demo_substitute":
        return ProductRouteFact(
            segment=segment,
            truth=ProductRouteTruth.DEMO_SUBSTITUTE,
            reason_id=ProductRouteReason.D047_DEMO_SUBSTITUTE_ACTIVE,
            evidence_ids=(ProductEvidenceId.D047_LEGACY_ROUTE_SELECTED,),
        )
    if legacy_route_class == "formal":
        if not formal_runtime_observed:
            return ProductRouteFact(
                segment=segment,
                truth=ProductRouteTruth.UNAVAILABLE,
                reason_id=(ProductRouteReason.FORMAL_ACTIVATION_EVIDENCE_MISSING),
                evidence_ids=(
                    ProductEvidenceId.GATE0_CONTRACT_ONLY,
                    ProductEvidenceId.NO_RUNTIME_EVIDENCE,
                ),
            )
        return ProductRouteFact(
            segment=segment,
            truth=ProductRouteTruth.FORMAL,
            reason_id=ProductRouteReason.FORMAL_ROUTE_OBSERVED,
            evidence_ids=formal_evidence_ids,
            formal_runtime_observed=True,
        )
    if legacy_route_class in {"unsupported", "unknown"}:
        return ProductRouteFact(
            segment=segment,
            truth=ProductRouteTruth.UNAVAILABLE,
            reason_id=ProductRouteReason.REQUESTED_ROUTE_UNAVAILABLE,
            evidence_ids=(
                ProductEvidenceId.GATE0_CONTRACT_ONLY,
                ProductEvidenceId.NO_RUNTIME_EVIDENCE,
            ),
        )
    raise ProductCompositionContractViolation("unknown legacy route class")

File: jiuwenswarm/channels/web/frontend/src/features/live-voice/formal/productCompositionContract.ts
export const PRODUCT_COMPOSITION_CONTRACT_VERSION = 'live-voice.product-composition.gate0.v1' as const;

export const PRODUCT_ROUTE_TRUTH = Object.freeze(['formal', 'fallback', 'demo_substitute', 'unavailable', 'disabled'] as const);
export type ProductRouteTruth = (typeof PRODUCT_ROUTE_TRUTH)[number];

export const PRODUCT_SEGMENTS = Object.freeze([
  'authority',
  'p1.speech_media',
  'p2.agent_interaction',
  'p3.query',
  'p3.control',
  'p3.progress',
  'browser.audio',
  'observability',
] as const);
export type ProductSegment = (typeof PRODUCT_SEGMENTS)[number];

export const PRODUCT_ROUTE_REASONS = Object.freeze([
  'FORMAL_ROUTE_OBSERVED',
  'EXPLICIT_FALLBACK_ACTIVE',
  'D047_DEMO_SUBSTITUTE_ACTIVE',
  'FEATURE_DISABLED',
  'ADAPTER_NOT_REGISTERED',
  'REQUESTED_ROUTE_UNAVAILABLE',
  'FORMAL_ACTIVATION_EVIDENCE_MISSING',
  'TRUSTED_AUTHORITY_UNAVAILABLE',
  'SPEECH_AUTHORIZATION_UNAVAILABLE',
  'MEDIA_AUTHORITY_UNAVAILABLE',
  'MEDIA_PROVIDER_UNAVAILABLE',
  'MEDIA_TRANSPORT_UNAVAILABLE',
  'MEDIA_LOGGER_ZERO_PERSISTENCE_UNPROVEN',
  'P2_AUTHORITY_UNAVAILABLE',
  'P2_RUNTIME_UNAVAILABLE',
  'P2_NOTIFICATION_BACKPRESSURE_UNRESOLVED',
  'P3_QUERY_AUTHORITY_UNAVAILABLE',
  'P3_CONFIRMATION_ISSUER_UNAVAILABLE',
  'TASK_PROGRESS_AUTHORITY_HANDOFF_UNAVAILABLE',
  'BROWSER_AUDIO_UNAVAILABLE',
  'OBSERVABILITY_CONSUMER_UNAVAILABLE',
  'ACTIVATION_FAILED',
] as const);
export type ProductRouteReason = (typeof PRODUCT_ROUTE_REASONS)[number];

export const PRODUCT_EVIDENCE_IDS = Object.freeze([
  'GATE0_CONTRACT_ONLY',
  'FEATURE_FLAG_OFF',
  'TRUSTED_AUTHORITY_RESOLVED',
  'FORMAL_ACTIVATION_LEASE_OPEN',
  'RUNTIME_PATH_OBSERVED',
  'P2_NOTIFICATION_BACKPRESSURE_CLOSED',
  'MEDIA_LOGGER_ZERO_PERSISTENCE_VERIFIED',
  'FALLBACK_ROUTE_SELECTED',
  'D047_LEGACY_ROUTE_SELECTED',
  'FORMAL_BATCH_SPEECH_FOUNDATION',
  'D059_AGENT_CR_FOUNDATION',
  'P3_QUERY_FOUNDATION',
  'TASK_EVENT_LIVE_ONLY',
  'PROGRESS_ARBITER_CONTIGUOUS_SEQUENCE_REQUIRED',
  'BROWSER_AUDIO_FOUNDATION',
  'OBSERVABILITY_FOUNDATION',
  'P2_NOTIFICATION_QUEUE_BLOCKING_RISK',
  'DEV_AUDIO_LOG_PERSISTENCE_RISK',
  'PREINTEGRATION_STATIC_AUDIT_ONLY',
  'PACKAGE_CONTRACT_ONLY',
  'NO_RUNTIME_EVIDENCE',
] as const);
export type ProductEvidenceId = (typeof PRODUCT_EVIDENCE_IDS)[number];

export type LegacyIntegratedRouteClass = 'formal' | 'fallback' | 'demo_substitute' | 'unsupported' | 'unknown';

export interface ProductRouteFact {
  readonly segment: ProductSegment;
  readonly truth: ProductRouteTruth;
  readonly reason_id: ProductRouteReason;
  readonly evidence_ids: readonly ProductEvidenceId[];
  readonly formal_runtime_observed: boolean;
}

export interface ProductCompositionManifest {
  readonly contract_version: typeof PRODUCT_COMPOSITION_CONTRACT_VERSION;
  readonly enabled: boolean;
  readonly routes: readonly ProductRouteFact[];
}

export class ProductCompositionContractViolation extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ProductCompositionContractViolation';
  }
}

const formalEvidence = new Set<ProductEvidenceId>(['TRUSTED_AUTHORITY_RESOLVED', 'FORMAL_ACTIVATION_LEASE_OPEN', 'RUNTIME_PATH_OBSERVED']);

const formalStopClosureBySegment: Readonly<Partial<Record<ProductSegment, ProductEvidenceId>>> = {
  'p1.speech_media': 'MEDIA_LOGGER_ZERO_PERSISTENCE_VERIFIED',
  'p2.agent_interaction': 'P2_NOTIFICATION_BACKPRESSURE_CLOSED',
};

const forbiddenEvidence: Readonly<Record<ProductRouteTruth, ReadonlySet<ProductEvidenceId>>> = {
  formal: new Set([
    'GATE0_CONTRACT_ONLY',
    'FEATURE_FLAG_OFF',
    'FALLBACK_ROUTE_SELECTED',
    'D047_LEGACY_ROUTE_SELECTED',
    'P2_NOTIFICATION_QUEUE_BLOCKING_RISK',
    'DEV_AUDIO_LOG_PERSISTENCE_RISK',
    'NO_RUNTIME_EVIDENCE',
  ]),
  fallback: new Set([...formalEvidence, 'FEATURE_FLAG_OFF', 'D047_LEGACY_ROUTE_SELECTED']),
  demo_substitute: new Set([...formalEvidence, 'FEATURE_FLAG_OFF', 'FALLBACK_ROUTE_SELECTED']),
  unavailable: new Set(['FEATURE_FLAG_OFF', 'FORMAL_ACTIVATION_LEASE_OPEN', 'RUNTIME_PATH_OBSERVED', 'FALLBACK_ROUTE_SELECTED', 'D047_LEGACY_ROUTE_SELECTED']),
  disabled: new Set(PRODUCT_EVIDENCE_IDS.filter(evidence => evidence !== 'FEATURE_FLAG_OFF')),
};

const allowedReasons: Readonly<Record<ProductRouteTruth, ReadonlySet<ProductRouteReason>>> = {
  formal: new Set(['FORMAL_ROUTE_OBSERVED']),
  fallback: new Set(['EXPLICIT_FALLBACK_ACTIVE']),
  demo_substitute: new Set(['D047_DEMO_SUBSTITUTE_ACTIVE']),
  disabled: new Set(['FEATURE_DISABLED']),
  unavailable: new Set(
    PRODUCT_ROUTE_REASONS.filter(
      reason =>
        reason !== 'FORMAL_ROUTE_OBSERVED' && reason !== 'EXPLICIT_FALLBACK_ACTIVE' && reason !== 'D047_DEMO_SUBSTITUTE_ACTIVE' && reason !== 'FEATURE_DISABLED'
    )
  ),
};

function freezeFact(input: ProductRouteFact): Readonly<ProductRouteFact> {
  if (!PRODUCT_SEGMENTS.includes(input.segment)) throw new ProductCompositionContractViolation('invalid product segment');
  if (!PRODUCT_ROUTE_TRUTH.includes(input.truth)) throw new ProductCompositionContractViolation('invalid product route truth');
  if (!allowedReasons[input.truth].has(input.reason_id)) {
    throw new ProductCompositionContractViolation('reason_id is incompatible with route truth');
  }
  if (!Array.isArray(input.evidence_ids) || input.evidence_ids.length === 0) {
    throw new ProductCompositionContractViolation('evidence_ids must be non-empty');
  }
  if (input.evidence_ids.some(evidence => !PRODUCT_EVIDENCE_IDS.includes(evidence))) {
    throw new ProductCompositionContractViolation('invalid product evidence id');
  }
  if (new Set(input.evidence_ids).size !== input.evidence_ids.length) {
    throw new ProductCompositionContractViolation('evidence_ids must not contain duplicates');
  }
  if (typeof input.formal_runtime_observed !== 'boolean') {
    throw new ProductCompositionContractViolation('formal_runtime_observed must be boolean');
  }
  const evidence = new Set(input.evidence_ids);
  if ([...forbiddenEvidence[input.truth]].some(id => evidence.has(id))) {
    throw new ProductCompositionContractViolation('route truth contains contradictory evidence');
  }
  if (input.truth === 'formal') {
    if (!input.formal_runtime_observed || [...formalEvidence].some(id => !evidence.has(id))) {
      throw new ProductCompositionContractViolation('formal truth requires trusted authority, an open activation lease, and observed runtime-path evidence');
    }
    const requiredStopClosure = formalStopClosureBySegment[input.segment];
    if (requiredStopClosure !== undefined && !evidence.has(requiredStopClosure)) {
      throw new ProductCompositionContractViolation('formal segment is missing its affirmative activation-stop closure evidence');
    }
  } else if (input.formal_runtime_observed) {
    throw new ProductCompositionContractViolation('non-formal truth cannot claim formal runtime observation');
  }
  const required: Partial<Record<ProductRouteTruth, ProductEvidenceId>> = {
    fallback: 'FALLBACK_ROUTE_SELECTED',
    demo_substitute: 'D047_LEGACY_ROUTE_SELECTED',
    disabled: 'FEATURE_FLAG_OFF',
  };
  if (required[input.truth] && !evidence.has(required[input.truth]!)) {
    throw new ProductCompositionContractViolation('route truth is missing its required evidence identifier');
  }
  return Object.freeze({ ...input, evidence_ids: Object.freeze([...input.evidence_ids]) });
}

function disabledFact(segment: ProductSegment): Readonly<ProductRouteFact> {
  return freezeFact({
    segment,
    truth: 'disabled',
    reason_id: 'FEATURE_DISABLED',
    evidence_ids: ['FEATURE_FLAG_OFF'],
    formal_runtime_observed: false,
  });
}

export function createProductCompositionManifest(input: {
  readonly enabled: boolean;
  readonly route_facts?: readonly ProductRouteFact[];
}): Readonly<ProductCompositionManifest> {
  if (typeof input.enabled !== 'boolean') throw new ProductCompositionContractViolation('enabled must be boolean');
  if (!input.enabled) {
    return Object.freeze({
      contract_version: PRODUCT_COMPOSITION_CONTRACT_VERSION,
      enabled: false,
      routes: Object.freeze(PRODUCT_SEGMENTS.map(disabledFact)),
    });
  }
  const routeFacts = input.route_facts ?? [];
  if (!Array.isArray(routeFacts)) {
    throw new ProductCompositionContractViolation('route_facts must be an array');
  }
  const facts = new Map<ProductSegment, Readonly<ProductRouteFact>>();
  for (const raw of routeFacts) {
    const fact = freezeFact(raw);
    if (facts.has(fact.segment)) throw new ProductCompositionContractViolation('duplicate product segment');
    facts.set(fact.segment, fact);
  }
  const routes = PRODUCT_SEGMENTS.map(
    segment =>
      facts.get(segment) ??
      freezeFact({
        segment,
        truth: 'unavailable',
        reason_id: 'ADAPTER_NOT_REGISTERED',
        evidence_ids: ['GATE0_CONTRACT_ONLY'],
        formal_runtime_observed: false,
      })
  );
  if (routes.slice(1).some(route => route.truth === 'formal') && routes[0].truth !== 'formal') {
    throw new ProductCompositionContractViolation('formal product segments require a formal authority segment');
  }
  return Object.freeze({
    contract_version: PRODUCT_COMPOSITION_CONTRACT_VERSION,
    enabled: true,
    routes: Object.freeze(routes),
  });
}

export function routeFactFromIntegratedShell(input: {
  readonly segment: ProductSegment;
  readonly feature_enabled: boolean;
  readonly legacy_route_class: unknown;
  readonly formal_runtime_observed?: boolean;
  readonly formal_evidence_ids?: readonly ProductEvidenceId[];
}): Readonly<ProductRouteFact> {
  if (typeof input.feature_enabled !== 'boolean') {
    throw new ProductCompositionContractViolation('feature_enabled must be boolean');
  }
  if (!input.feature_enabled) return disabledFact(input.segment);
  if (!PRODUCT_SEGMENTS.includes(input.segment)) throw new ProductCompositionContractViolation('invalid product segment');
  if (input.formal_runtime_observed !== undefined && typeof input.formal_runtime_observed !== 'boolean') {
    throw new ProductCompositionContractViolation('formal_runtime_observed must be boolean');
  }
  if (typeof input.legacy_route_class !== 'string') {
    throw new ProductCompositionContractViolation('legacy_route_class must be string');
  }
  if (input.legacy_route_class === 'fallback') {
    return freezeFact({
      segment: input.segment,
      truth: 'fallback',
      reason_id: 'EXPLICIT_FALLBACK_ACTIVE',
      evidence_ids: ['FALLBACK_ROUTE_SELECTED'],
      formal_runtime_observed: false,
    });
  }
  if (input.legacy_route_class === 'demo_substitute') {
    return freezeFact({
      segment: input.segment,
      truth: 'demo_substitute',
      reason_id: 'D047_DEMO_SUBSTITUTE_ACTIVE',
      evidence_ids: ['D047_LEGACY_ROUTE_SELECTED'],
      formal_runtime_observed: false,
    });
  }
  if (input.legacy_route_class === 'formal') {
    if (input.formal_runtime_observed !== true) {
      return freezeFact({
        segment: input.segment,
        truth: 'unavailable',
        reason_id: 'FORMAL_ACTIVATION_EVIDENCE_MISSING',
        evidence_ids: ['GATE0_CONTRACT_ONLY', 'NO_RUNTIME_EVIDENCE'],
        formal_runtime_observed: false,
      });
    }
    return freezeFact({
      segment: input.segment,
      truth: 'formal',
      reason_id: 'FORMAL_ROUTE_OBSERVED',
      evidence_ids: input.formal_evidence_ids ?? [],
      formal_runtime_observed: true,
    });
  }
  if (input.legacy_route_class === 'unsupported' || input.legacy_route_class === 'unknown') {
    return freezeFact({
      segment: input.segment,
      truth: 'unavailable',
      reason_id: 'REQUESTED_ROUTE_UNAVAILABLE',
      evidence_ids: ['GATE0_CONTRACT_ONLY', 'NO_RUNTIME_EVIDENCE'],
      formal_runtime_observed: false,
    });
  }
  throw new ProductCompositionContractViolation('unknown legacy route class');
}
